staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go

```diff
@@ -175,6 +183,7 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
 		return apierrors.NewInternalError(err)
 	}
 	attr.VersionedObject = newVersionedObject
+	attr.Attributes.GetReinvocationState().MarkMutated()
```

Comment:
A single bit doesn't tell us where in the chain mutation occurred. I think we need more information to correctly re-invoke:
* we don't want to reinvoke if all mutations occurred prior to the first call to a webhook
* we do want to reinvoke if a mutation occurred in a later webhook
* we do want to reinvoke if a mutation occurred in a reinvocation of an in-tree plugin

I think we need these bits:
```
// lastOutput holds the result of the last MutatingAdmissionWebhook#Admit call
lastOutput runtime.Object
// reinvokeWebhooksIfNeeded holds the set of webhooks to reinvoke if a later mutation occurs
reinvokeWebhooksIfNeeded sets.String
// reinvokeWebhooks holds the set of webhooks that should be reinvoked
reinvokeWebhooks sets.String
// reinvokeInTree indicates in-tree plugins should be reinvoked
reinvokeInTree bool
```

1. When we evaluate a mutating webhook, if it has an `IfNeeded` reinvocation policy, add it to the `reinvokeWebhooksIfNeeded` set
2. When a mutating webhook results in a mutation:
  1. add the current content of `reinvokeWebhooksIfNeeded` to `reinvokeWebhooks`
  2. clear `reinvokeWebhooksIfNeeded`
  3. set `reinvokeInTree` to true
3. When MutatingAdmissionWebhook#Admit completes, save a copy of attr.GetObject() as `lastOutput`
4. When MutatingAdmissionWebhook#Admit is reinvoked, if `!deepEqual(lastOutput, attr.GetObject())`:
  1. add the current content of `reinvokeWebhooksIfNeeded` to `reinvokeWebhooks`
  2. clear `reinvokeWebhooksIfNeeded`
5. When considering whether to reinvoke a webhook, check if it is in the `reinvokeWebhooks` set

Reply:
We'll want test cases around these scenarios:

in-tree -> mutation
webhook -> no mutation
complete (no reinvocation needed)

in-tree -> mutation
webhook -> mutation
reinvoke in-tree -> no mutation
complete (in-tree didn't change anything, so no webhook reinvocation required)

in-tree -> mutation
webhook -> mutation
reinvoke in-tree -> mutation
reinvoke webhook -> mutation
complete (both reinvoked once, no additional loops)

in-tree -> mutation
webhook A -> mutation
webhook B -> mutation
reinvoke in-tree -> no mutation
reinvoke webhook A -> no mutation
complete (no mutations occurred after webhook B was called, no reinvocation required)

in-tree -> mutation
webhook A -> mutation
webhook B -> mutation
reinvoke in-tree -> no mutation
reinvoke webhook A -> mutation
reinvoke webhook B -> mutation (webhook A mutated after webhook B was called, one reinvocation required)
complete (both reinvoked once, no additional loops)

Reply:
Thanks @liggitt I'm implementing this. The test cases are hugely helpful. I'll send out a update (in a separate commit to this PR) tonight.

Reply:
@liggitt I feel like there is still more testing I'd like to do, but I've completed a first pass on the implementation and tests.